framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,433 @@
+package org.checkerframework.common.value;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+
+/**
+ * The Range Class with mathematics operations. Models the range indicated by the @IntRange
+ * annotation.
+ *
+ * <p>Note that this class is not responsible for detecting incorrect parameters, e.g. the value of
+ * "from" could be greater than the value of "to". This incorrectness would be eventually caught by
+ * {@link
+ * org.checkerframework.common.value.ValueAnnotatedTypeFactory#createIntRangeAnnotation(Range)} when
+ * creating an annotation from range, which would then be replaced with @UnknownVal.
+ *
+ * @author JasonMrX
+ */
+public class Range {
+
+    /** The value 'from' */
+    public final long from;
+
+    /** The value 'to' */
+    public final long to;
+
+    /**
+     * Constructs a range with its bounds specified by two parameters, "from" and "to".
+     *
+     * <p>Note that it is possible to construct a range with incorrect parameters, e.g. the value of
+     * "from" could be greater than the value to "to". This incorrectness would be caught by {@link
+     * org.checkerframework.common.value.ValueAnnotatedTypeFactory#createIntRangeAnnotation(Range)}
```

Comment:
1.  Below is a way you could implement the `NOTHING` constant.  The `Range.java` code would never call the special constructor except to create the `NOTHING` constant.  After testing (e.g., `right == NOTHING`), then the code can assume that the Range is well-formed.  An alternate representation would add an extra `boolean` field to the representation, but I don't think that is necessary.

2.  `NOTHING` represents that at run time, there are no possible values; that is, the expression is dead code and is never executed.  Any operation on `NOTHING` and something else is still `NOTHING`.  This is another way of saying what you said yourself:  a transfer function, when applied to bottom, outputs bottom.

```java
    /** A range containing all possible values. */
    public final EVERYTHING = new Range(Long.MIN_VALUE, Long.MAX_VALUE);

    /** The empty range. */
    public final NOTHING = new Range(true);

    /**
     * Constructs a non-empty range with its bounds specified by two parameters, "from" and "to".
     *
     * @param from the lower bound (inclusive)
     * @param to the higher bound (inclusive)
     */
    public Range(long from, long to) {
        if (!(from <= to)) {
            throw new IllegalArgumentException(String.format("Invalid Range: %s %s", from, to));
        }
        this.from = from;
        this.to = to;
    }

    /** Creates the singleton empty range. */
    private Range(boolean isEmpty) {
        if (! isEmpty) {
            throw new IllegalArgumentException();
        }
        this.from = Long.MAX_VALUE;
        this.to = Long.MIN_VALUE;
    }
```